src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -1373,6 +1375,537 @@ private Autogenerated.StateItem ToAutogeneratedStateItem(StateTransactionRequest
             return new UnsubscribeConfigurationResponse(resp.Ok, resp.Message);
         }
+
+        #endregion
+
+        #region Cryptography
+
+        private async Task<byte[]> EncryptStreamAsync(string vaultResourceName, Stream stream,
+            KeyWrapAlgorithm algorithm, string keyName, DataEncryptionCipher dataEncryptionCipher,
+            string decryptionKeyName, bool omitDecryptionKey, CancellationToken cancellationToken = default)
+        {
+            ArgumentVerifier.ThrowIfNullOrEmpty(vaultResourceName, nameof(vaultResourceName));
+            ArgumentVerifier.ThrowIfNullOrEmpty(keyName, nameof(keyName));
+
+
+            var encryptRequestOptions = new Autogenerated.EncryptRequestOptions
+            {
+                ComponentName = vaultResourceName,
+                DataEncryptionCipher = dataEncryptionCipher.GetValueFromEnumMember(),
+                KeyName = keyName,
+                KeyWrapAlgorithm = algorithm.GetValueFromEnumMember(),
+                OmitDecryptionKeyName = omitDecryptionKey
+            };
+
+            if (!omitDecryptionKey)
+            {
+                ArgumentVerifier.ThrowIfNullOrEmpty(decryptionKeyName, nameof(decryptionKeyName));
+                encryptRequestOptions.DecryptionKeyName = decryptionKeyName;
+            }
+
+            var options = CreateCallOptions(headers: null, cancellationToken);
+            var duplexStream = client.EncryptAlpha1(options);
+
+            //Start with passing the metadata about the encryption request itself in the first message
+            await duplexStream.RequestStream.WriteAsync(
+                new Autogenerated.EncryptRequest { Options = encryptRequestOptions }, cancellationToken);
+
+            //Send the plaintext bytes in 4 kb blocks in subsequent messages
+            const int blockSize = 4 * 1024; // 4 kb
+            await using (var bufferedStream = new BufferedStream(stream, blockSize))
+            {
+                var buffer = new byte[blockSize];
+                int bytesRead;
+                ulong sequenceNumber = 0;
+
+                while ((bytesRead = await bufferedStream.ReadAsync(buffer, 0, blockSize, cancellationToken)) != 0)
+                {
+                    await duplexStream.RequestStream.WriteAsync(new Autogenerated.EncryptRequest
+                    {
+                        Payload = new Autogenerated.StreamPayload
+                        {
+                            Data = ByteString.CopyFrom(buffer, 0, bytesRead), Seq = sequenceNumber
+                        }
+                    }, cancellationToken);
+
+                    //Increment the sequence number
+                    sequenceNumber++;
+                }
+            }
+            
+            //Send the completion message
+            await duplexStream.RequestStream.CompleteAsync();
+
+            //Set up how the response is handled since we know it's of finite length
+            var returnedBytes = new List<byte>();
```

Comment:
Is this significantly different than writing to a `MemoryStream` and then returning its underlying buffer?

Reply:
Either one results in an allocation via `ToArray()` ultimately and a MemoryStream would have a slightly higher overhead than a list. As either is just a buffer for the bytes, I think there's little value in swapping over to a MemoryStream, but it was a worthwhile learning experience to give it a shot.